### PR TITLE
output variable for cycling ratio (for unitary system)

### PR DESCRIPTION
**Pull request by JanghyunJK**
Pull request overview
---------------------

- This is coming from EUSS HP-RTU upsizing measure work (anticipated data release date = 2024 November).

- When we (including @ChristopherCaradonna) initially got reviews from Paul T. and Landan T., the concern in upsizing heat pumps was mostly around increased short cycling of heat pump.

- So, we want to add output variables that can capture short cycling metric. Before, we have operating COP values that can sort of give us the impact of short cycling but that is not a direct output var that only captures the impact of short cycling.

- Short cycling is calculated with two values per building model:
  1. hourly timeseries output vars are enabled: `Unitary System DX Coil Cycling Ratio`, `Unitary System Sensible Cooling Rate`, `Unitary System Sensible Heating Rate`.
  2. timeseries cycling ratio for cooling operation (=output var 1) is calculated by filtering cycling ratio values only when cooling rate is non-zero.
  3. timeseries cycling ratio for heating operation (=output var 2) is calculated by filtering cycling ratio values only when heating rate is non-zero.
  4. get (non-weighted) average (so no more timeseries) of cooling cycling ratio and heating cycling ratio.
  5. if there are multiple unitary systems in a building model, then get weighted average (by floor area served) cycling ratios.

- unsure if I changed all necessary stuffs to enable this.

- I have not filled in everything in this file since I wasn't sure what the columns were asking: `Variable List.xlsx`

- I might want to see if the output vars are really added in comstock wide csv with some small comstock run.

### Pull Request Author

This pull request makes changes to (select all the apply):
 - [ ] Documentation
 - [ ] Infrastructure (includes apptainer image, buildstock batch, dependencies, continuous integration tests)
 - [ ] Sampling
 - [ ] Workflow Measures
 - [ ] Upgrade Measures
 - [x] Reporting Measures
 - [x] Postprocessing

Author pull request checklist:

 - [x] Tagged the pull request with the appropriate label (documentation, infrastructure, sampling, workflow measure, upgrade measure, reporting measure, postprocessing) to help categorize changes in the release notes.
 - [ ] Added tests for new measures
 - [x] Updated measure .xml(s)
 - [x] Register values added to `comstock_column_definitions.csv`
 - [ ] Both `options_lookup.tsv` files updated
 - [ ] 10k+ test run
 - [ ] Change documentation written
 - [ ] Measure documentation written
 - [ ] ComStock documentation updated
 - [ ] Changes reflected in example `.yml` files
 - [ ] Changes reflected in `README.md` files
 - [ ] Added 'See ComStock License' language to first two lines of each code file
 - [ ] Implements corresponding measure tests and indexing path in `test/measure_tests.txt` or/and `test/resource_measure_tests.txt`
 - [ ] All new and existing tests pass the CI

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a code review on GitHub
 - [ ] All related changes have been implemented: data and method additions, changes, tests
 - [ ] If fixing a defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] Reviewed change documentation
 - [ ] Ensured code files contain License reference
 - [ ] Results differences are reasonable
 - [ ] Make sure the newly added measures has been added with tests and indexed properly
 - [ ] CI status: all tests pass

### CODE REVIEW

**Comment by JanghyunJK** [comment]
@ChristopherCaradonna confirmed!

image

File: measures/comstock_sensitivity_reports/measure.rb [hunk 1]
```diff
@@ -2460,6 +2486,39 @@ def run(runner, user_arguments)
...
+      ts_unitary_sys_sen_heating_rate = sql.timeSeries(ann_env_pd, timeseries_timestep, 'Unitary System Sensible Heating Rate', airloopunisys.name.to_s.upcase) # W
+      ts_unitary_sys_sen_heating_rate = convert_timeseries_to_list(ts_unitary_sys_sen_heating_rate)
+      # calculate average cycling ratio
+      ts_unitary_sys_cycling_ratio_cooling = ts_unitary_sys_cycling_ratio.zip(ts_unitary_sys_sen_cooling_rate).select { |a, b| b != 0 }.map { |a, b| a }
+      avg_unitary_sys_cycling_ratio_cooling = get_average_from_array(ts_unitary_sys_cycling_ratio_cooling)
+      ts_unitary_sys_cycling_ratio_heating = ts_unitary_sys_cycling_ratio.zip(ts_unitary_sys_sen_heating_rate).select { |a, b| b != 0 }.map { |a, b| a }
+      avg_unitary_sys_cycling_ratio_heating = get_average_from_array(ts_unitary_sys_cycling_ratio_heating)
+      # get floor area served
+      corresponding_thermal_zone = airloopunisys.thermalZone.get
+      corresponding_floor_area_m_2 = corresponding_thermal_zone.floorArea
```
Comment (ChristopherCaradonna) [inline measures/comstock_sensitivity_reports/measure.rb]: I don't know if floor area is best way to average. I think either by heating/cooling energy, or heating/cooling capacity would be better.

File: measures/comstock_sensitivity_reports/measure.rb [hunk 2]
```diff
@@ -179,6 +184,9 @@ def energyPlusOutputRequests(runner, user_arguments)
     result << OpenStudio::IdfObject.load("Output:Variable,*,Water Heater Propane Energy,RunPeriod;").get # J
     result << OpenStudio::IdfObject.load('Output:Variable,*,Water Heater Heating Energy,RunPeriod;').get # J
     result << OpenStudio::IdfObject.load('Output:Variable,*,Water Heater Unmet Demand Heat Transfer Energy,RunPeriod;').get # J
+    result << OpenStudio::IdfObject.load("Output:Variable,*,Unitary System DX Coil Cycling Ratio,#{timeseries_timestep};").get # -
+    result << OpenStudio::IdfObject.load("Output:Variable,*,Unitary System Sensible Cooling Rate,#{timeseries_timestep};").get # W
```
Comment (ChristopherCaradonna) [inline measures/comstock_sensitivity_reports/measure.rb]: I am not as familiar with these unitary system variables. If we are just using it to determine if there is heating or cooling, then maybe it's fine. But if we are using it for the calculation, I would prefer using the heating and cooling rate from the coil objects.

**Review by ChristopherCaradonna — COMMENTED** [review]
Looks good, I added some comments below. Let's test on a 10k run before we approve and merge. Have you tested on a multispeed coil?

File: measures/comstock_sensitivity_reports/measure.rb (hunk 1 above)
Comment (JanghyunJK) [reply measures/comstock_sensitivity_reports/measure.rb]: yeah I can take a look on this one.

File: measures/comstock_sensitivity_reports/measure.rb (hunk 2 above)
Comment (JanghyunJK) [reply measures/comstock_sensitivity_reports/measure.rb:188]: @ChristopherCaradonna that may cause: building with 30 pszs ending up with 60 (30 for cooling and 30 for heating) timeseries (each with 8760 datapoints) for that specific model, maybe? burden on simulation workflow? I'll check what the document says for the variables above. Let me know if my example with 60 timeseries output vars seems really what's gonna happen.

File: measures/comstock_sensitivity_reports/measure.rb (hunk 1 above)
Comment (JanghyunJK) [reply measures/comstock_sensitivity_reports/measure.rb]: [c5222de](https://github.com/NREL/ComStock/pull/190/commits/c5222de47ae1e13be5ffa51f08cbb77e05187292)

**Comment by JanghyunJK** [comment]
updates:

- made some more changes so now the four new output variables are,
  - weighted average cycling ratio for cooling (one value per model)
  - weighted average cycling ratio for heating (one value per model)
  - weighted average percentage of excessive use of electricity (for cooling) due to cycling (one value per model)
  - weighted average percentage of excessive use of electricity (for heating) due to cycling (one value per model)

- because not all coil objects have cycling ratio output var, this calculation only works for
  - two speed coils
  - and multi-speed coils
  - and if coil type is other than the two above, it will report zero for all four output vars

- moving on to 10k testing

**Comment by JanghyunJK** [comment]
merging to HPRTU PR: https://github.com/NREL/ComStock/pull/163

**Pull request merged**